src/store.ts

```diff
@@ -0,0 +1,269 @@
+import Gio from "gi://Gio";
+import GLib from "gi://GLib";
+
+import { StickyError, StickyErrorType } from "./errors.js";
+import { INote, Note } from "./util.js";
+
+Gio._promisify(
+  Gio.File.prototype,
+  "load_contents_async",
+  "load_contents_finish",
+);
+
+Gio._promisify(
+  Gio.File.prototype,
+  "replace_contents_async",
+  "replace_contents_finish",
+);
+
+const decoder = new TextDecoder();
+const encoder = new TextEncoder();
+
+interface SavedNoteData {
+  v: 1;
+  notes: INote[];
+  state: State;
+}
+
+interface LoadNotesReturn {
+  notes: Note[];
+  state: State;
+}
+
+interface State {
+  all_notes: boolean;
+}
+
+export const NotesDir = Gio.file_new_for_path(
+  GLib.build_filenamev([GLib.get_user_data_dir(), pkg.name]),
+);
+
+export const NewNotesDir = Gio.file_new_for_path(
+  GLib.build_filenamev([GLib.get_user_data_dir(), pkg.name, "notes"]),
+);
+
+export const Notes = Gio.file_new_for_path(
+  GLib.build_filenamev([NotesDir.get_path()!, "notes.json"]),
+);
+
+export const load_file = (file: Gio.File) => {
+  try {
+    Notes.get_parent()!.make_directory_with_parents(null);
+  } catch (e: unknown) {
+    if (e instanceof GLib.Error) {
+      if (!e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.EXISTS)) {
+        throw new StickyError(
+          StickyErrorType.NO_PERMISSION,
+          "Failed to create notes directory",
+        );
+      }
+    }
+  }
+
+  return file.load_contents_async(null)
+    .then(([contents]) => {
+      try {
+        return JSON.parse(decoder.decode(contents));
+      } catch (e: unknown) {
+        throw new StickyError(
+          StickyErrorType.FILE_CORRUPTED,
+          "Failed to parse notes",
+        );
+      }
+    }).catch((e: unknown) => {
+      if (e instanceof GLib.Error) {
+        if (e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.NOT_FOUND)) {
+          throw new StickyError(
+            StickyErrorType.FILE_NOT_FOUND,
+            "Failed to load notes",
+          );
+        } else {
+          throw new StickyError(
+            StickyErrorType.UNKNOWN,
+            "Failed to load notes",
+          );
+        }
+      }
+    });
+};
+
+export const save_file = (file: Gio.File, data: any) => {
+  const string = JSON.stringify(data, null, 2);
+  // const etag = GLib.compute_checksum_for_string(
+  //   GLib.ChecksumType.MD5,
+  //   string,
+  //   string.length,
+  // );
+
+  return file.replace_contents_async(
+    encoder.encode(string),
+    null,
+    false,
```

Comment:
> If make_backup is TRUE, this function will attempt to make a backup of file. Internally, it uses g_file_replace(), so will try to replace the file contents in the safest way possible. For example, atomic renames are used when replacing local files’ contents.

https://docs.gtk.org/gio/method.File.replace_contents.html

it's a but unclear from the documentation if that means `make_backup` should be set to `true` for "the safest way possible".

Could you try and see if the backup file persists?

Reply:
you mean I should set the "make_backup" option to true?

what do you mean by "see if the backup file persists"

Reply:
I meant to see if `make_backup` keeps the backup file after `replace_contents` is finished.

Reply:
I think note (pun intended)...